cpp/src/kpca/kpca.cuh

```diff
@@ -0,0 +1,350 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#pragma once
+
+#include <common/device_utils.cuh>
+
+#include <cuml/common/utils.hpp>
+#include <cuml/decomposition/params.hpp>
+#include <cuml/matrix/kernelparams.h>
+
+#include <raft/core/handle.hpp>
+#include <raft/distance/kernels.cuh>
+#include <raft/linalg/detail/cublas_wrappers.hpp>
+#include <raft/linalg/eig.cuh>
+#include <raft/linalg/gemm.cuh>
+#include <raft/linalg/subtract.cuh>
+#include <raft/matrix/math.cuh>
+#include <raft/matrix/matrix.cuh>
+#include <raft/stats/mean.cuh>
+#include <raft/stats/mean_center.cuh>
+#include <raft/util/cuda_utils.cuh>
+
+#include <rmm/device_uvector.hpp>
+#include <rmm/device_vector.hpp>
+#include <rmm/exec_policy.hpp>
+
+#include <cuda_runtime.h>
+#include <thrust/device_ptr.h>
+#include <thrust/fill.h>
+#include <thrust/find.h>
+#include <thrust/reduce.h>
+
+#include <tsvd/tsvd.cuh>
+
+#include <utility>
+
+namespace ML {
+
+// All the positive eigenvalues that are too small (with a value smaller than the maximum
+// eigenvalue multiplied by for double precision 1e-12 (2e-7 for float)) are set to zero.
```

Comment:
When removing zero eigen values, we treat small values as zero. See reference from sklearn: https://github.com/scikit-learn/scikit-learn/blob/b72e81af473c079ae95314efbca86557a836defa/sklearn/utils/validation.py#L1901-L1903